## Cleaning Data 

Sprocket Central Pty Ltd , a medium size bikes & cycling accessories organisation, has approached Tony Smith (Partner) in KPMG’s Lighthouse & Innovation Team. Sprocket Central Pty Ltd  is keen to learn more about KPMG’s expertise in its Analytics, Information & Modelling team. 

The client provided KPMG with 3 datasets:

- Customer Demographic 

- Customer Addresses

- Transactions data in the past 3 months

##### My mission is to check the datasets for any mistakes and try to get a high quality datasets

# Starting point

## Load the datasets

Get the data and load them into a dataframe and prepare them for analyzing

In [1]:
#import the required Libraries
import pandas as pd
import numpy as np

In [2]:
#load the datasets into dataframe

address = pd.read_csv('addresses.csv')
print("The first 5 rows of the dataframe") 
address.head()

The first 5 rows of the dataframe


customer_id              address  postcode            state    country  \
0            1   060 Morning Avenue      2016  New South Wales  Australia   
1            2  6 Meadow Vale Court      2153  New South Wales  Australia   
2            4   0 Holy Cross Court      4211              QLD  Australia   
3            5  17979 Del Mar Point      2448  New South Wales  Australia   
4            6     9 Oakridge Court      3216              VIC  Australia   

   property_valuation  
0                  10  
1                  10  
2                   9  
3                   4  
4                   9

In [3]:
print("The last 5 rows of the dataframe") 

address.tail()

The last 5 rows of the dataframe


customer_id                    address  postcode state    country  \
3994         3999            1482 Hauk Trail      3064   VIC  Australia   
3995         4000  57042 Village Green Point      4511   QLD  Australia   
3996         4001     87 Crescent Oaks Alley      2756   NSW  Australia   
3997         4002           8194 Lien Street      4032   QLD  Australia   
3998         4003            320 Acker Drive      2251   NSW  Australia   

      property_valuation  
3994                   3  
3995                   6  
3996                  10  
3997                   7  
3998                   7

In [4]:
#check the transaction with describe method

address.describe(include='all')

customer_id                  address     postcode state    country  \
count   3999.000000                     3999  3999.000000  3999       3999   
unique          NaN                     3996          NaN     5          1   
top             NaN  3 Mariners Cove Terrace          NaN   NSW  Australia   
freq            NaN                        2          NaN  2054       3999   
mean    2003.987997                      NaN  2985.755939   NaN        NaN   
std     1154.576912                      NaN   844.878364   NaN        NaN   
min        1.000000                      NaN  2000.000000   NaN        NaN   
25%     1004.500000                      NaN  2200.000000   NaN        NaN   
50%     2004.000000                      NaN  2768.000000   NaN        NaN   
75%     3003.500000                      NaN  3750.000000   NaN        NaN   
max     4003.000000                      NaN  4883.000000   NaN        NaN   

        property_valuation  
count          3999.000000  
unique                 NaN  
top                    NaN  
freq                   NaN  
mean              7.514379  
std               2.824663  
min               1.000000  
25%               6.000000  
50%               8.000000  
75%              10.000000  
max              12.000000

In [5]:
# Find the Empty values or NULL values
missing_data = address.isnull()
for column in missing_data.columns.values.tolist():
    print(column)
    print (missing_data[column].value_counts())
    print("")   

customer_id
False    3999
Name: customer_id, dtype: int64

address
False    3999
Name: address, dtype: int64

postcode
False    3999
Name: postcode, dtype: int64

state
False    3999
Name: state, dtype: int64

country
False    3999
Name: country, dtype: int64

property_valuation
False    3999
Name: property_valuation, dtype: int64



### Analyzing Address Dataset results

Using describe, head and tail methods we got:
- at The customer_id column, The last customer id is 4003 but the number of the available data is 3999, that mean 4 are missing
- at state column, name of state is not standard and different for same state, for example NSW and New South Wales where they refer to the same.
- Data type of columns (country, state, address) need to be changed


#### Solving all issues at address dataset

In [6]:
# at The customer_id column

In [7]:
# at state column

In [8]:
# correcting data types

## Next dataset

In [9]:
#load the datasets into dataframe
print("The first 5 rows of the dataframe") 
trans = pd.read_csv('transactions.csv')
trans.head()

The first 5 rows of the dataframe


transaction_id  product_id  customer_id transaction_date online_order  \
0               1           2         2950       25-02-2017        False   
1               2           3         3120       21-05-2017         True   
2               3          37          402       16-10-2017        False   
3               4          88         3135       31-08-2017        False   
4               5          78          787       01-10-2017         True   

  order_status           brand product_line product_class product_size  \
0     Approved           Solex     Standard        medium       medium   
1     Approved   Trek Bicycles     Standard        medium        large   
2     Approved      OHM Cycles     Standard           low       medium   
3     Approved  Norco Bicycles     Standard        medium       medium   
4     Approved  Giant Bicycles     Standard        medium        large   

   list_price standard_cost  product_first_sold_date  
0       71.49        $53.62                  41245.0  
1     2091.47       $388.92                  41701.0  
2     1793.43       $248.82                  36361.0  
3     1198.46       $381.10                  36145.0  
4     1765.30       $709.48                  42226.0

In [10]:
print("The last 5 rows of the dataframe") 
trans.tail()

The last 5 rows of the dataframe


transaction_id  product_id  customer_id transaction_date online_order  \
19995           19996          51         1018       24-06-2017         True   
19996           19997          41          127       09-11-2017         True   
19997           19998          87         2284       14-04-2017         True   
19998           19999           6         2764       03-07-2017        False   
19999           20000          11         1144       22-09-2017         True   

      order_status          brand product_line product_class product_size  \
19995     Approved     OHM Cycles     Standard          high       medium   
19996     Approved          Solex         Road        medium       medium   
19997     Approved     OHM Cycles     Standard        medium       medium   
19998     Approved     OHM Cycles     Standard          high       medium   
19999     Approved  Trek Bicycles     Standard        medium        small   

       list_price standard_cost  product_first_sold_date  
19995     2005.66     $1,203.40                  37823.0  
19996      416.98       $312.74                  35560.0  
19997     1636.90        $44.71                  40410.0  
19998      227.88       $136.73                  38216.0  
19999     1775.81     $1,580.47                  36334.0

In [11]:
#check the transaction with describe method
trans.describe(include='all')

transaction_id   product_id   customer_id transaction_date  \
count     20000.000000  20000.00000  20000.000000            20000   
unique             NaN          NaN           NaN              364   
top                NaN          NaN           NaN       14-02-2017   
freq               NaN          NaN           NaN               82   
mean      10000.500000     45.36465   1738.246050              NaN   
std        5773.647028     30.75359   1011.951046              NaN   
min           1.000000      0.00000      1.000000              NaN   
25%        5000.750000     18.00000    857.750000              NaN   
50%       10000.500000     44.00000   1736.000000              NaN   
75%       15000.250000     72.00000   2613.000000              NaN   
max       20000.000000    100.00000   5034.000000              NaN   

       online_order order_status  brand product_line product_class  \
count         19640        20000  19803        19803         19803   
unique            2            2      6            4             3   
top            True     Approved  Solex     Standard        medium   
freq           9829        19821   4253        14176         13826   
mean            NaN          NaN    NaN          NaN           NaN   
std             NaN          NaN    NaN          NaN           NaN   
min             NaN          NaN    NaN          NaN           NaN   
25%             NaN          NaN    NaN          NaN           NaN   
50%             NaN          NaN    NaN          NaN           NaN   
75%             NaN          NaN    NaN          NaN           NaN   
max             NaN          NaN    NaN          NaN           NaN   

       product_size    list_price standard_cost  product_first_sold_date  
count         19803  20000.000000         19803             19803.000000  
unique            3           NaN           103                      NaN  
top          medium           NaN       $388.92                      NaN  
freq          12990           NaN           465                      NaN  
mean            NaN   1107.829449           NaN             38199.776549  
std             NaN    582.825242           NaN              2875.201110  
min             NaN     12.010000           NaN             33259.000000  
25%             NaN    575.270000           NaN             35667.000000  
50%             NaN   1163.890000           NaN             38216.000000  
75%             NaN   1635.300000           NaN             40672.000000  
max             NaN   2091.470000           NaN             42710.000000

In [12]:
# Find the Empty values or NULL values
missing_data = trans.isnull()
for column in missing_data.columns.values.tolist():
    print(column)
    print (missing_data[column].value_counts())
    print("")   

transaction_id
False    20000
Name: transaction_id, dtype: int64

product_id
False    20000
Name: product_id, dtype: int64

customer_id
False    20000
Name: customer_id, dtype: int64

transaction_date
False    20000
Name: transaction_date, dtype: int64

online_order
False    19640
True       360
Name: online_order, dtype: int64

order_status
False    20000
Name: order_status, dtype: int64

brand
False    19803
True       197
Name: brand, dtype: int64

product_line
False    19803
True       197
Name: product_line, dtype: int64

product_class
False    19803
True       197
Name: product_class, dtype: int64

product_size
False    19803
True       197
Name: product_size, dtype: int64

list_price
False    20000
Name: list_price, dtype: int64

standard_cost
False    19803
True       197
Name: standard_cost, dtype: int64

product_first_sold_date
False    19803
True       197
Name: product_first_sold_date, dtype: int64



### Analyzing Transaction Dataset results

From Transaction dataset, by describe, head and tail methods we got:
- Empty fields are found at the columns (online_order : 360 , brand : 197 , product_line , product_class : 197 , product_size : 197 , standard_cost : 197 , product_first_sold_date : 197)
- It is better to change the repeated values at columns (online_order, order_status, product_line, product_class,  product_size) into a numeric value to make it easier to evaluate and analyze.
- data type of columns (transaction_date, online_order, order_status, brand, product_line, product_class, product_size, standard_cost)


#### Solving all issues at Transaction dataset

In [13]:
# droping or changing the missing data

In [14]:
# create new numeric columns for (online_order, order_status, product_line, product_class,  product_size)

In [15]:
# correct the data types


## Next

In [16]:
#load the datasets into dataframe
print("The first 5 rows of the dataframe") 
customers = pd.read_csv('CustomersWithID.csv')
customers.head()

The first 5 rows of the dataframe


customer_id      first_name  last_name  gender  \
0            1         Laraine  Medendorp       F   
1            2             Eli    Bockman    Male   
2            3           Arlin     Dearle    Male   
3            4          Talbot        NaN    Male   
4            5  Sheila-kathryn     Calton  Female   

   past_3_years_bike_related_purchases         DOB               job_title  \
0                                   93  12-10-1953     Executive Secretary   
1                                   81  16-12-1980  Administrative Officer   
2                                   61  20-01-1954      Recruiting Manager   
3                                   33  03-10-1961                     NaN   
4                                   56  13-05-1977           Senior Editor   

  job_industry_category     wealth_segment deceased_indicator owns_car  tenure  
0                Health      Mass Customer                  N      Yes    11.0  
1    Financial Services      Mass Customer                  N      Yes    16.0  
2              Property      Mass Customer                  N      Yes    15.0  
3                    IT      Mass Customer                  N       No     7.0  
4                   NaN  Affluent Customer                  N      Yes     8.0

In [17]:
print("The last 5 rows of the dataframe") 
customers.tail()

The last 5 rows of the dataframe


customer_id first_name last_name  gender  \
3995         3996    Rosalia  Halgarth  Female   
3996         3997     Blanch    Nisuis  Female   
3997         3998     Sarene   Woolley       U   
3998         3999  Patrizius       NaN    Male   
3999         4000      Kippy   Oldland    Male   

      past_3_years_bike_related_purchases         DOB              job_title  \
3995                                    8  09-08-1975  VP Product Management   
3996                                   87  13-07-2001        Statistician II   
3997                                   60         NaN      Assistant Manager   
3998                                   11  24-10-1973                    NaN   
3999                                   76  05-11-1991   Software Engineer IV   

     job_industry_category     wealth_segment deceased_indicator owns_car  \
3995                Health      Mass Customer                  N       No   
3996         Manufacturing     High Net Worth                  N      Yes   
3997                    IT     High Net Worth                  N       No   
3998         Manufacturing  Affluent Customer                  N      Yes   
3999                   NaN  Affluent Customer                  N       No   

      tenure  
3995    19.0  
3996     1.0  
3997     NaN  
3998    10.0  
3999    11.0

In [18]:
#check the transaction with describe method

customers.describe(include='all')

customer_id first_name last_name  gender  \
count   4000.000000       4000      3875    4000   
unique          NaN       3139      3725       6   
top             NaN        Max  Pristnor  Female   
freq            NaN          5         3    2037   
mean    2000.500000        NaN       NaN     NaN   
std     1154.844867        NaN       NaN     NaN   
min        1.000000        NaN       NaN     NaN   
25%     1000.750000        NaN       NaN     NaN   
50%     2000.500000        NaN       NaN     NaN   
75%     3000.250000        NaN       NaN     NaN   
max     4000.000000        NaN       NaN     NaN   

        past_3_years_bike_related_purchases         DOB  \
count                           4000.000000        3913   
unique                                  NaN        3448   
top                                     NaN  30-01-1978   
freq                                    NaN           7   
mean                              48.890000         NaN   
std                               28.715005         NaN   
min                                0.000000         NaN   
25%                               24.000000         NaN   
50%                               48.000000         NaN   
75%                               73.000000         NaN   
max                               99.000000         NaN   

                                   job_title job_industry_category  \
count                                   3494                  3344   
unique                                   195                     9   
top     Business Systems Development Analyst         Manufacturing   
freq                                      45                   799   
mean                                     NaN                   NaN   
std                                      NaN                   NaN   
min                                      NaN                   NaN   
25%                                      NaN                   NaN   
50%                                      NaN                   NaN   
75%                                      NaN                   NaN   
max                                      NaN                   NaN   

       wealth_segment deceased_indicator owns_car       tenure  
count            4000               4000     4000  3913.000000  
unique              3                  2        2          NaN  
top     Mass Customer                  N      Yes          NaN  
freq             2000               3998     2024          NaN  
mean              NaN                NaN      NaN    10.657041  
std               NaN                NaN      NaN     5.660146  
min               NaN                NaN      NaN     1.000000  
25%               NaN                NaN      NaN     6.000000  
50%               NaN                NaN      NaN    11.000000  
75%               NaN                NaN      NaN    15.000000  
max               NaN                NaN      NaN    22.000000

In [19]:
# Find the Empty values or NULL values
missing_data = customers.isnull()
for column in missing_data.columns.values.tolist():
    print(column)
    print (missing_data[column].value_counts())
    print("")   

customer_id
False    4000
Name: customer_id, dtype: int64

first_name
False    4000
Name: first_name, dtype: int64

last_name
False    3875
True      125
Name: last_name, dtype: int64

gender
False    4000
Name: gender, dtype: int64

past_3_years_bike_related_purchases
False    4000
Name: past_3_years_bike_related_purchases, dtype: int64

DOB
False    3913
True       87
Name: DOB, dtype: int64

job_title
False    3494
True      506
Name: job_title, dtype: int64

job_industry_category
False    3344
True      656
Name: job_industry_category, dtype: int64

wealth_segment
False    4000
Name: wealth_segment, dtype: int64

deceased_indicator
False    4000
Name: deceased_indicator, dtype: int64

owns_car
False    4000
Name: owns_car, dtype: int64

tenure
False    3913
True       87
Name: tenure, dtype: int64



### Analyzing Customers Dataset results

From Customers DataSet, It has:
- The column default contains many wrong and dangerous entries, when entering data there should be, standardized encoding, check dangerous entries like data injections or JavaScript attacks, also harmful values that exceed the capacity of memory or processing.
- The security aspects are very important to consider during data entry to eliminate any danger move that could destroy the dataset.
- Empty or missing data at columns (last name : 125 , DOB : 87 , job_title : 506 , job_industry_category : 656 , tenure : 87)
- wrong data types for columns (first_name, last_name, gender, DOB, job_title, job_industry_category, wealth_segment, deceased_indicator, owns_car)
- change the columns with repeated values to numeric values analyze them easily.
- The DOB impossible values there should be a specific range of age.

#### Solving all issues at Customers dataset

In [20]:
# Drop the default Column

In [21]:
# droping or changing the missing data

In [22]:
# create new numeric columns for (gender, job_industry_category, wealth_segment, deceased_indicator, owns_car)

In [23]:
# Correct the data types